Makefile

```
@@ -36,98 +59,166 @@ else
+# tools
+PRINT = printf
 
-# Versions J and below used the C matrix math implementations
-MGU_MATRIX_FILES := mtxcatf normalize scale translate
-ifeq ($(MGU_ASM), 1)
-C_FILES := $(filter-out $(addprefix src/gu/,$(MGU_MATRIX_FILES:=.c)),$(C_FILES))
-else
-S_FILES := $(filter-out $(addprefix src/mgu/,$(MGU_MATRIX_FILES:=.s)),$(S_FILES))
+ifeq ($(COLOR),1)
+NO_COL  := \033[0m
+RED     := \033[0;31m
+GREEN   := \033[0;32m
+BLUE    := \033[0;34m
+YELLOW  := \033[0;33m
+BLINK   := \033[33;5m
 endif
 
-C_O_FILES := $(foreach f,$(C_FILES:.c=.o),$(BUILD_DIR)/$f)
-S_O_FILES := $(foreach f,$(S_FILES:.s=.o),$(BUILD_DIR)/$f)
-O_FILES   := $(S_O_FILES) $(C_O_FILES)
+# Common build print status function
+define print
+  @$(PRINT) "$(GREEN)$(1) $(YELLOW)$(2)$(GREEN) -> $(BLUE)$(3)$(NO_COL)\n"
+endef
 
-$(shell mkdir -p src $(foreach dir,$(SRC_DIRS),$(BUILD_DIR)/$(dir)))
+#==============================================================================#
+# Main Targets                                                                 #
+#==============================================================================#
 
-.PHONY: all clean distclean setup
-all: $(BUILD_AR)
-
-$(BUILD_AR): $(O_FILES)
-	@printf "    [AR] $@\n"
-	$(V)$(AR) rcs $@ $^
+# Default target
+default: $(LIB)
 
 clean:
-	$(RM) -rf $(BUILD_DIR)
+	$(RM) -r $(BUILD_DIR_BASE)
 
-distclean:
-	$(RM) -rf extracted/ $(BUILD_ROOT)
+ALL_DIRS := $(BUILD_DIR) $(addprefix $(BUILD_DIR)/,$(SRC_DIRS))
 
-GBIDEFINE := -DF3DEX_GBI
+# Make sure build directory exists before compiling anything
+DUMMY != mkdir -p $(ALL_DIRS)
 
+$(BUILD_DIR)/src/voice/%.o: CFLAGS += -I$(WORKING_DIR)/src/voice
+$(BUILD_DIR)/src/voice/%.o: DEFINES += LANG_JAPANESE=1
 $(BUILD_DIR)/src/gu/parse_gbi.o: GBIDEFINE := -DF3D_GBI
 $(BUILD_DIR)/src/gu/us2dex_emu.o: GBIDEFINE :=
 $(BUILD_DIR)/src/gu/us2dex2_emu.o: GBIDEFINE :=
 $(BUILD_DIR)/src/sp/sprite.o: GBIDEFINE := -DF3D_GBI
 $(BUILD_DIR)/src/sp/spriteex.o: GBIDEFINE :=
 $(BUILD_DIR)/src/sp/spriteex2.o: GBIDEFINE :=
-$(BUILD_DIR)/src/voice/%.o: OPTFLAGS += -DLANG_JAPANESE -I$(WORKING_DIR)/src -I$(WORKING_DIR)/src/voice
-$(BUILD_DIR)/src/voice/%.o: CC := $(WORKING_DIR)/tools/compile_sjis.py -D__CC=$(CC) -D__BUILD_DIR=$(BUILD_DIR)
+
+#==============================================================================#
+# Compilation Recipes                                                          #
+#==============================================================================#
+
+# Compile C code
+$(BUILD_DIR)/src/voice/%.o: src/voice/%.c
+	$(call print,Compiling:,$<,$@)
+	$(V)tools/compile_sjis.py -D__CC=$(CC) -D__BUILD_DIR=$(BUILD_DIR) -c $(CFLAGS) -MMD -MF $(BUILD_DIR)/src/voice/$*.d  -o $@ $<
 
 $(BUILD_DIR)/%.o: %.c
-	@printf "    [CC] $<\n"
-	$(V)$(CC) $(CFLAGS) $(MIPS_VERSION) $(CPPFLAGS) $(OPTFLAGS) $< $(IINC) -o $@
-	$(V)tools/set_o32abi_bit.py $@
+	$(call print,Compiling:,$<,$@)
+	$(V)$(CC) -c $(CFLAGS) $(GBIDEFINE) -MMD -MF $(BUILD_DIR)/$*.d  -o $@ $<
```

```suggestion
	$(V)$(CC) -c $(CFLAGS) $(GBIDEFINE) -MD -MMD -MP  -o $@ $<
```
The default .d file outpath path on any version of gcc worth using is to sit them next to the .o file, there shouldn't be a need to specify it here

---

Fixed according to suggestion